autoblocks/_impl/tracerv2/run.py

```diff
@@ -0,0 +1,263 @@
+started: bool = False
+
+
+# Context
+current_test_id_var = contextvars.ContextVar("current_test_id")
+current_test_case_var = contextvars.ContextVar("current_test_case")
+
+
+def get_or_create_ingestion_client(ingestion_key: str) -> httpx.Client:
+    global global_ingestion_client
+    if global_ingestion_client is None:
+        global_ingestion_client = httpx.Client(
+            headers={"Authorization": f"Bearer {ingestion_key}"},
+        )
+    return global_ingestion_client
+
+
+def run_event_loop(_loop: asyncio.AbstractEventLoop) -> None:
+    asyncio.set_event_loop(_loop)
+    _loop.run_forever()
+
+
+async def gather_with_max_concurrency(
+    max_concurrency: int,
+    coroutines: List[Coroutine],
+) -> None:
+    """
+    Borrowed from https://stackoverflow.com/a/61478547
+    """
+    semaphore = asyncio.Semaphore(max_concurrency)
+
+    async def sem_coro(coro: Coroutine):
+        async with semaphore:
+            return await coro
+
+    # return_exceptions=True causes exceptions to be returned as values instead
+    # of propagating them to the caller. this is similar in behavior to Promise.allSettled
+    await asyncio.gather(*(sem_coro(c) for c in coroutines), return_exceptions=True)
+
+
+class AutoblocksTracer:
+    def __init__(
+        self,
+        ingestion_key: Optional[str] = None,
+        # Initialize the tracer with a trace_id. All events sent with this tracer will
+        # send this trace ID unless overwritten by:
+        # - calling set_trace_id
+        # - specifying a trace_id when calling send_event
+        trace_id: Optional[str] = None,
+        # Initialize the tracer with properties. All events sent with this tracer will
+        # send these properties unless overwritten by:
+        # - calling set_properties
+        # Additionally, these properties can be updated by:
+        # - calling update_properties
+        # - specifying properties when calling send_event
+        properties: Optional[Dict] = None,
+        # Timeout for sending events to Autoblocks
+        timeout: timedelta = timedelta(seconds=5),
+    ):
+        self._trace_id: Optional[str] = trace_id
+        self._properties: Dict = properties or {}
+
+        ingestion_key = ingestion_key or AutoblocksEnvVar.INGESTION_KEY.get()
+        if not ingestion_key:
+            raise ValueError(
+                f"You must provide an ingestion_key or set the {AutoblocksEnvVar.INGESTION_KEY} environment variable."
+            )
+
+        self._client = get_or_create_ingestion_client(ingestion_key)
+        self._timeout_seconds = timeout.total_seconds()
+        cli_server_address = AutoblocksEnvVar.CLI_SERVER_ADDRESS.get()
+        if not cli_server_address:
+            raise RuntimeError(
+                "Autoblocks tests must be run within the context of the testing CLI.\n"
+                "Make sure you are running your test command with:\n"
+                "$ npx autoblocks testing exec -- <your test command>"
+            )
```

Comment:
can drop this, only relevant for testing sdk

Reply:
Copypasta 👍